### PR TITLE
Lazy load room messages

### PR DESCRIPTION
**Pull request by ismailgulek**
Part of vector-im/element-ios#4382

There are some methods causing a room's messages to be loaded. They are:
- [ ] `MXRoom.refreshOutgoingMessages`
- [ ] `MXRoom.sentStatus (sentStatusWithCompletion newly)`

### CODE REVIEW

File: MatrixSDK/Data/MXRoomSummary.m [hunk 1]
```diff
@@ -887,7 +889,6 @@ - (NSUInteger)hash
 
     result = prime * result + [_lastMessage.eventId hash];
     result = prime * result + [_lastMessage.text hash];
-    result = prime * result + self.room.storedMessagesCount;
```
Comment (ismailgulek) [inline MatrixSDK/Data/MXRoomSummary.m:890]: We'll need to test this.

File: MatrixSDK/Data/Store/MXFileStore/MXFileStore.m [hunk 2]
```diff
@@ -823,9 +827,42 @@ - (MXMemoryRoomStore*)getOrCreateRoomStore:(NSString*)roomId
...
-        roomStores[roomId] = roomStore;
+        NSString *roomFile = [self messagesFileForRoom:roomId forBackup:NO];
+        
+        if ([[NSFileManager defaultManager] fileExistsAtPath:roomFile])
+        {
+            MXFileRoomStore *roomStore;
+            @try
+            {
+                NSDate *startDate = [NSDate date];
+                roomStore = [NSKeyedUnarchiver unarchiveObjectWithFile:roomFile];
```
Comment (manuroe) [inline MatrixSDK/Data/Store/MXFileStore/MXFileStore.m:838]: It should be processed on a separated thread.

File: MatrixSDK/VoIP/MXCallManager.m [hunk 3]
```diff
@@ -1732,65 +1732,80 @@ - (void)directCallableRoomWithVirtualUser:(NSString * _Nonnull)userId
...
+    dispatch_group_t dispatchGroup = dispatch_group_create();
     for (MXRoom *room in rooms) {
-        id<MXEventsEnumerator> enumerator = [room enumeratorForStoredMessagesWithTypeIn:@[kMXEventTypeStringCallInvite]];
-        MXEvent *callEvent = enumerator.nextEvent;
-        if (callEvent)
-        {
-            [callEvents addObject:callEvent];
-        }
+        dispatch_group_enter(dispatchGroup);
+        [room enumeratorForStoredMessagesWithTypeIn:@[kMXEventTypeStringCallInvite]
```
Comment (manuroe) [inline MatrixSDK/VoIP/MXCallManager.m:1762]: I can see now a bug on this part of the code. This is not due to the changes to this PR.
`enumeratorForStoredMessagesWithTypeIn` returns only raw content of events. It does not decrypt them. It cannot see message type.

**Review by manuroe — COMMENTED** [review]
This feature probably requires more architecture changes. And perhaps, a decoupling between MXFileRoomStore and MXMemoryRoomStore.

To implement this feature correctly, we have to make any access to current MXMemoryRoomStore.messages and MXMemoryRoomStore.messagesByEventIds. But those objects are supposed to exist after the init MXMemoryRoomStore.

File: MatrixSDK/Data/MXEventTimeline.m [hunk 4]
```diff
@@ -156,6 +156,23 @@ - (void)destroy
     }
 }
 
+- (void)storeMessagesEnumerator:(void (^)(id<MXEventsEnumerator>))completion
+{
+    if (cachedStoreMessagesEnumerator)
+    {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            completion(self->cachedStoreMessagesEnumerator);
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply MatrixSDK/Data/MXEventTimeline.m:164]: It should be nullable, updated.

File: MatrixSDK/Data/MXEventTimeline.m [hunk 5]
```diff
@@ -194,7 +212,8 @@ - (void)resetPagination
     backState = [[MXRoomState alloc] initBackStateWith:_state];
 
     // Reset store pagination
-    storeMessagesEnumerator = [store messagesEnumeratorForRoom:_state.roomId];
+    cachedStoreMessagesEnumerator = nil;
+    [self storeMessagesEnumerator:nil];
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply MatrixSDK/Data/MXEventTimeline.m:216]: Thanks, removed.

File: MatrixSDK/Data/MXRoomSummary.m (hunk 1 above)
Comment (ismailgulek) [reply MatrixSDK/Data/MXRoomSummary.m:890]: It also works without the number of stored messages

File: MatrixSDK/Data/Store/MXFileStore/MXFileStore.m (hunk 2 above)
Comment (ismailgulek) [reply MatrixSDK/Data/Store/MXFileStore/MXFileStore.m]: It is now.

File: MatrixSDK/MXSession.m [hunk 6]
```diff
@@ -2837,30 +2837,36 @@ - (MXHTTPOperation*)eventWithEventId:(NSString*)eventId
     {
         // Try to find it from the store first
         // (this operation requires a roomId for the moment)
-        MXEvent *event = [_store eventWithEventId:eventId inRoom:roomId];
+        __block MXEvent *event = [_store eventWithEventId:eventId inRoom:roomId];
```
[the remark this answers is not in the document]
Comment (ismailgulek) [reply MatrixSDK/MXSession.m]: It is now.

**Comment by ismailgulek** [comment]
Replaced by https://github.com/matrix-org/matrix-ios-sdk/pull/1206

**Pull request closed without merging**